### PR TITLE
Standardize crypto errors

### PR DESCRIPTION
**Pull request by chaitanyab2311**
# Description

Enriched errors in cryptography API. This is still a draft as unit tests need to be updated and integration tests need to be added

## Issue reference

Please reference the issue this PR will close: #7491 

## Checklist

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing

### CODE REVIEW

**Comment by chaitanyab2311** [comment]
@cicoyle I am getting some errors related to MockActors in http_test while running tests locally. Am i missing something

File: pkg/api/errors/cryptography.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	ErrCryptoEncryptOperation      = "ENCRYPT"
+	ErrCryptoDecryptOperation      = "DECRYPT"
+	ErrCryptoKeyOperation          = "KEY"
+	ErrCryptoVerifySignature       = "VERIFY_SIGNATURE"
+)
+
+const ResourceType = "Cryptography"
+
+func CryptoNotConfigured(name string) error {
+	message := "crypto providers not configured"
```
Comment (ItalyPaleAle) [inline pkg/api/errors/cryptography.go:39]: Please use consts for these things

File: pkg/api/errors/cryptography.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		message,
+		"ERR_CRYPTO_PROVIDER_GET_KEY",
+	).
+		WithErrorInfo(kiterrors.CodePrefixCryptography+ErrCryptoGetKey, nil).
+		WithResourceInfo("", name, "", message).
+		Build()
+}
+
+func CryptoProviderEncryptOperation(name string, err error) error {
+	message := fmt.Sprintf("failed to perform encrypt operation: %s", err.Error())
```
Comment (ItalyPaleAle) [inline pkg/api/errors/cryptography.go:91]:
```suggestion
	message := "failed to perform encrypt operation: " + err.Error()
```

File: pkg/api/errors/cryptography.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		message,
+		"ERR_CRYPTO_ENCRYPT_OPERATION",
+	).
+		WithErrorInfo(kiterrors.CodePrefixBindings+ErrCryptoEncryptOperation, nil).
+		WithResourceInfo("", name, "", message).
+		Build()
+}
+
+func CryptoProviderDecryptOperation(name string, err error) error {
+	message := fmt.Sprintf("failed to perform decrypt operation: %s", err.Error())
```
Comment (ItalyPaleAle) [inline pkg/api/errors/cryptography.go:104]:
```suggestion
	message := "failed to perform decrypt operation: " + err.Error()
```

File: pkg/api/errors/cryptography.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		http.StatusInternalServerError,
+		message,
+		"ERR_CRYPTO_DECRYPT_OPERATION",
+	).
+		WithErrorInfo(kiterrors.CodePrefixBindings+ErrCryptoDecryptOperation, nil).
+		WithResourceInfo("", name, "", message).
+		Build()
+}
+
+func CryptoProviderKeyOperation(name string, err string) error {
```
Comment (ItalyPaleAle) [inline pkg/api/errors/cryptography.go:116]: Err should be an error, also for consistency

File: pkg/api/errors/cryptography.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		http.StatusInternalServerError,
+		message,
+		"ERR_CRYPTO_KEY_OPERATION",
+	).
+		WithErrorInfo(kiterrors.CodePrefixBindings+ErrCryptoKeyOperation, nil).
+		WithResourceInfo("", name, "", message).
+		Build()
+}
+
+func CryptoProviderVerifySignatureOperation(name string, err string) error {
```
Comment (ItalyPaleAle) [inline pkg/api/errors/cryptography.go:129]: Ditto

File: pkg/api/universal/subtlecrypto_subtlecrypto.go [hunk 6]
```diff
@@ -233,7 +233,8 @@ func (a *Universal) SubtleWrapKeyAlpha1(ctx context.Context, in *runtimev1pb.Sub
 		// We are not going to return the exact error from the component to the user, because an error that is too specific could allow for various side channel attacks (e.g. AES-CBC and padding oracle attacks)
 		// We will log the full error as a debug log, but only return a generic one to the user
 		a.logger.Debug(messages.ErrCryptoOperation.WithFormat(err))
-		err = messages.ErrCryptoOperation.WithFormat("failed to wrap key")
```
Comment (ItalyPaleAle) [inline pkg/api/universal/subtlecrypto_subtlecrypto.go:236]: Very important: please do not return the exact error returned by the component! That can be logged, but must not be returned to the user. The old code was overriding the error for a reason.

See the comment above. The exact error must **never** be sent to users, or really bad things (side channel attacks) could happen

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

**Comment by ItalyPaleAle** [comment]
Please note that with crypto operations, errors are sensitive as they can disclose information to users that can be used for attacks (side channel).

Please make sure errors from the crypto components are never returned to callers. They are ok to log, but not returned in the API response - return a generic "failed to perform operation" error instead.

**Comment by cicoyle** [comment]
> @cicoyle I am getting some errors related to MockActors in http_test while running tests locally. Am i missing something

I believe you might need the tags in your Go Tool Args. When you run the tests in Dapr you will likely need to add these tags: `-tags=stablecomponents,unit`
image

Once you get to the integration tests too, you will need a tag: `go test -v -race -tags integration ./tests/integration -focus errors`, where `errors` might change pending what you'd like to 'focus' on

File: pkg/api/errors/cryptography.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+		http.StatusInternalServerError,
+		message,
+		"ERR_CRYPTO_PROVIDERS_NOT_CONFIGURED",
+	).
+		WithErrorInfo(kiterrors.CodePrefixCryptography+ErrCryptoProviderNotConfigured, nil).
+		WithResourceInfo(ResourceType, name, "", message).
+		Build()
+}
+
+func CryptoNotFound(name string) error {
```
Comment (cicoyle) [inline pkg/api/errors/cryptography.go:51]: same here. There is a generic `NotFound` you can use here: https://github.com/dapr/dapr/blob/master/pkg/api/errors/errors.go#L11

File: pkg/api/errors/cryptography.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+		http.StatusInternalServerError,
+		message,
+		"ERR_CRYPTO_MISSING_COMPONENT_NAME",
+	).
+		WithErrorInfo(kiterrors.CodePrefixCryptography+ErrCryptoNameEmpty, nil).
+		WithResourceInfo(ResourceType, name, "", message).
+		Build()
+}
+
+func CryptoProviderGetKey(name string, err string) error {
```
Comment (cicoyle) [inline pkg/api/errors/cryptography.go:77]:
```suggestion
func CryptoProviderGetKey(name string, msg string) error {
```
Then you can get rid of the `message := err` line below :)

Or if the intent was to use the string version of the initial error, then it should be of type `error` when passed in

File: pkg/api/errors/cryptography.go (hunk 1 above)
Comment (chaitanyab2311) [reply pkg/api/errors/cryptography.go]: Added

File: pkg/api/universal/subtlecrypto_subtlecrypto.go (hunk 6 above)
Comment (chaitanyab2311) [reply pkg/api/universal/subtlecrypto_subtlecrypto.go:236]: Thanks for this. I should have read the comments properly. Modified the changes accordingly

File: pkg/api/universal/subtlecrypto_subtlecrypto.go [hunk 9]
```diff
@@ -159,7 +163,8 @@ func (a *Universal) SubtleEncryptAlpha1(ctx context.Context, in *runtimev1pb.Sub
 		// We are not going to return the exact error from the component to the user, because an error that is too specific could allow for various side channel attacks (e.g. AES-CBC and padding oracle attacks)
 		// We will log the full error as a debug log, but only return a generic one to the user
 		a.logger.Debug(messages.ErrCryptoOperation.WithFormat(err))
-		err = messages.ErrCryptoOperation.WithFormat("failed to encrypt")
+		err = fmt.Errorf("failed to encrypt")
+		err = apierrors.CryptoProviderEncryptOperation(in.ComponentName, err)
```
Comment (ItalyPaleAle) [inline pkg/api/universal/subtlecrypto_subtlecrypto.go:167]: Nit: use [`errors.New`](https://pkg.go.dev/errors#New) if there's no format (like, no `%s`), which is faster (same below)

```suggestion
		err = apierrors.CryptoProviderEncryptOperation(in.ComponentName, errors.New("failed to encrypt"))
```

File: pkg/api/universal/subtlecrypto_subtlecrypto.go [hunk 10]
```diff
@@ -191,7 +196,8 @@ func (a *Universal) SubtleDecryptAlpha1(ctx context.Context, in *runtimev1pb.Sub
 		// We are not going to return the exact error from the component to the user, because an error that is too specific could allow for various side channel attacks (e.g. AES-CBC and padding oracle attacks)
 		// We will log the full error as a debug log, but only return a generic one to the user
 		a.logger.Debug(messages.ErrCryptoOperation.WithFormat(err))
-		err = messages.ErrCryptoOperation.WithFormat("failed to decrypt")
+		err = fmt.Errorf("failed to decrypt")
+		err = apierrors.CryptoProviderDecryptOperation(in.ComponentName, err)
```
Comment (ItalyPaleAle) [inline pkg/api/universal/subtlecrypto_subtlecrypto.go:200]:
```suggestion
		err = apierrors.CryptoProviderDecryptOperation(in.ComponentName, errors.New("failed to decrypt"))
```

File: pkg/api/universal/subtlecrypto_subtlecrypto.go [hunk 11]
```diff
@@ -233,7 +239,8 @@ func (a *Universal) SubtleWrapKeyAlpha1(ctx context.Context, in *runtimev1pb.Sub
 		// We are not going to return the exact error from the component to the user, because an error that is too specific could allow for various side channel attacks (e.g. AES-CBC and padding oracle attacks)
 		// We will log the full error as a debug log, but only return a generic one to the user
 		a.logger.Debug(messages.ErrCryptoOperation.WithFormat(err))
-		err = messages.ErrCryptoOperation.WithFormat("failed to wrap key")
+		err = fmt.Errorf("failed to wrap key")
+		err = apierrors.CryptoProviderKeyOperation(in.ComponentName, err)
```
Comment (ItalyPaleAle) [inline pkg/api/universal/subtlecrypto_subtlecrypto.go:243]:
```suggestion
		err = apierrors.CryptoProviderKeyOperation(in.ComponentName, errors.New("failed to wrap key"))
```

File: pkg/api/universal/subtlecrypto_subtlecrypto.go [hunk 12]
```diff
@@ -265,7 +272,8 @@ func (a *Universal) SubtleUnwrapKeyAlpha1(ctx context.Context, in *runtimev1pb.S
 		// We are not going to return the exact error from the component to the user, because an error that is too specific could allow for various side channel attacks (e.g. AES-CBC and padding oracle attacks)
 		// We will log the full error as a debug log, but only return a generic one to the user
 		a.logger.Debug(messages.ErrCryptoOperation.WithFormat(err))
-		err = messages.ErrCryptoOperation.WithFormat("failed to unwrap key")
+		err = fmt.Errorf("failed to unwrap key")
+		err = apierrors.CryptoProviderKeyOperation(in.ComponentName, err)
```
Comment (ItalyPaleAle) [inline pkg/api/universal/subtlecrypto_subtlecrypto.go:276]:
```suggestion
		err = apierrors.CryptoProviderKeyOperation(in.ComponentName, errors.New("failed to unwrap key"))
```

File: pkg/api/universal/subtlecrypto_subtlecrypto.go [hunk 13]
```diff
@@ -306,7 +314,8 @@ func (a *Universal) SubtleSignAlpha1(ctx context.Context, in *runtimev1pb.Subtle
 		// We are not going to return the exact error from the component to the user, because an error that is too specific could allow for various side channel attacks (e.g. AES-CBC and padding oracle attacks)
 		// We will log the full error as a debug log, but only return a generic one to the user
 		a.logger.Debug(messages.ErrCryptoOperation.WithFormat(err))
-		err = messages.ErrCryptoOperation.WithFormat("failed to sign")
+		err = fmt.Errorf("failed to sign")
+		err = apierrors.CryptoProviderKeyOperation(in.ComponentName, err)
```
Comment (ItalyPaleAle) [inline pkg/api/universal/subtlecrypto_subtlecrypto.go:318]:
```suggestion
		err = apierrors.CryptoProviderKeyOperation(in.ComponentName, errors.New("failed to sign"))
```

File: pkg/api/universal/subtlecrypto_subtlecrypto.go [hunk 14]
```diff
@@ -337,7 +346,8 @@ func (a *Universal) SubtleVerifyAlpha1(ctx context.Context, in *runtimev1pb.Subt
 		// We are not going to return the exact error from the component to the user, because an error that is too specific could allow for various side channel attacks (e.g. AES-CBC and padding oracle attacks)
 		// We will log the full error as a debug log, but only return a generic one to the user
 		a.logger.Debug(messages.ErrCryptoOperation.WithFormat(err))
-		err = messages.ErrCryptoOperation.WithFormat("failed to verify signature")
+		err = fmt.Errorf("failed to verify signature: ")
+		err = apierrors.CryptoProviderVerifySignatureOperation(in.ComponentName, err)
```
Comment (ItalyPaleAle) [inline pkg/api/universal/subtlecrypto_subtlecrypto.go:350]:
```suggestion
		err = apierrors.CryptoProviderVerifySignatureOperation(in.ComponentName, errors.New("failed to verify signature"))
```

File: pkg/api/errors/cryptography.go
Comment (ItalyPaleAle) [inline pkg/api/errors/cryptography.go:1]: Just a nit, I'd name this "crypto.go" since that's how all other files are named

**Comment by yaron2** [comment]
@chaitanyab2311 are you planning to take this PR out of draft?

**Comment by daixiang0** [comment]
ping @chaitanyab2311

**Pull request closed without merging**